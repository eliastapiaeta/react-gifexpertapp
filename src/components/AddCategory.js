import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategorias }) => {

    const [inputValue, setinputValue] = useState('');
    const [inputNumberValue, setInputNumberValue] = useState(10);

    const handleInputChangeText = (e) => {
        const valorFiltrado = numtext(e.target.value);
        const valorLimitado = limLengthStr(valorFiltrado, 25);

        if (valorLimitado) {
            setinputValue(valorFiltrado);
            // console.log('inputValue: ' + inputValue);
            // console.log('valorFiltrado: ' + valorFiltrado);
        }
    }
    const handleInputChangeNum = (e) => {
        const cantidad = 2;
        const inValue = (e.target.value !== "") && e.target.value;
        const valorFiltrado = stringToNum(inValue, cantidad);
        const { valorLimitado } = (valorFiltrado) && limLengthStr(valorFiltrado.toString(), cantidad);

        // console.log('\n');
        // console.log('inValue: ' + inValue);
        // console.log('valorFiltrado: ' + valorFiltrado);
        // console.log('valorLimitado: ' + valorLimitado);
        // console.log('valorAlcanzado: ' + valorAlcanzado);
        // console.log('\n');
        if (valorLimitado) {
            setInputNumberValue(valorFiltrado);
            // console.log('inputValue: ' + inputValue);
            // console.log('valorFiltrado: ' + valorFiltrado);
        }
        if (!valorFiltrado) {
            setInputNumberValue('');
        }
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        // console.log('handleSubmit', inputValue);

        if (inputValue.trim().length > 2) {
            const inputNum = (inputNumberValue !== "") && inputNumberValue;

            // console.log('inputNum: ' + inputNum);

            if (inputNum) {
                // console.log('submit OK');
                // console.log('inputValue ', inputValue, 'typeof inputValue', typeof inputValue);
                // console.log('inputNumberValue ', inputNumberValue, 'typeof inputNumberValue', typeof inputNumberValue);

                setCategorias(cats => [{ num: inputNumberValue, cat: inputValue }, ...cats]);
                setinputValue('');
            }
        }

    }
    const stringToNum = (valNaN, cantidad) => {

        const valNull = (valNaN !== null) && valNaN;
        const regex = /(\d+)/g;

        // console.log('valNaN: ' + valNaN + ' typeof: ' + typeof valNaN);
        // console.log('valNull: ' + valNull + ' typeof: ' + typeof valNull);

        const valString = (typeof valNull !== 'string') ? valNull.toString() : valNull;

        // console.log('valString: ' + valString + ' typeof: ' + typeof valString);

        const num = valString ? valString.match(regex) : null;
        const haveNumber = (num !== null) && num;

        // console.log('num:::: ' + typeof haveNumber + ' ' + haveNumber);

        const resulString = (haveNumber) && (haveNumber.toString()).substring(0, cantidad);
        const resulNumber = Number(resulString);

        return resulNumber;
    }
    const numtext = (val) => {
        let out = '';
        //Se añaden las letras validas
        let filtro = ' abcdefghijklmnñopqrstuvwxyzABCDEFGHIJKLMNÑOPQRSTUVWXYZ1234567890';//Caracteres validos

        for (let i = 0; i < val.length; i++)
            if (filtro.indexOf(val.charAt(i)) !== -1)
                out += val.charAt(i);
        return out;
    }
    const limLengthStr = (valor, limite) => {
        // console.log(limite);
        const vmin = 0;
        const vmax = valor.length;
        const valorAlcanzado = (limite === vmax) && true;
        // console.log('\n valor: ' + valor);
        // console.log('vmin: ' + vmin + ' vmax: ' + vmax);
        const valorLimitado = ((vmax >= vmin) && (vmax <= limite)) && true;
        return { valorLimitado, valorAlcanzado };
    }

    return (
        <form onSubmit={(e) => handleSubmit(e)}>
            <div id='formGif'>
                <div id='textGifs'>
                    <label>
                        Imagen
                        <input
                            id="inText"
                            placeholder='Imagen a buscar'
                            type='text'
                            maxLength='25'
                            pattern="[A-Za-z0-9 ]{1,30}"
                            value={inputValue}
                            onChange={handleInputChangeText}
                        />
                    </label>
                </div>
                <div id='numberGifs'>
                    <label>
                        Lote
                        <input
                            id='inNum'
                            placeholder='n°'
                            type='text'
                            maxLength='3'
                            pattern='[0-9]{1,2}'
                            value={inputNumberValue}
                            onChange={handleInputChangeNum}
                        />
                    </label>
                </div>
                <div id='divEnter'>
                    <label>
                        <br />
                        <input type="submit" id="btnForm" value="Ok" />
                    </label>
                </div>
                <br id='brGifs' />
            </div>
        </form>
    )
}

AddCategory.propTypes = {
    setCategorias: PropTypes.func.isRequired
}
