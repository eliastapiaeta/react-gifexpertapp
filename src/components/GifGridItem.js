import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ({ id, title, url }) => {
    // console.log(id, title, url);
    return (
        <div className='card animate__animated animate__fadeInLeft'>
            <img id={id} src={url} alt={title} />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    id : PropTypes.string.isRequired,
    title : PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}

/**
 * 1. Enzymme
 * 2. Enzyme to Json
 * 3. Debe mostrar el componente correctamente
 *      * shallow
 *      * wrapper
 *      * wrapper .toMatchSnapshot()
 */
