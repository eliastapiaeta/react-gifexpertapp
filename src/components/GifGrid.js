import React, { /* useState, useEffect */ } from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';


export const GifGrid = ({ categoria, quantity }) => {

    // console.log(categoria, quantity);
    const { data: images, loading } = useFetchGifs(categoria, quantity);

    return (
        <>
            <h3 className='animate__animated animate__bounceIn'>{categoria} ({quantity})</h3>

            { /* loading ? 'Cargando...' : 'Data cargada' */}
            {loading && <p className='animate__animated animate__flash'>Loading...</p>}

            {<div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                        // <li key={id}>{title}</li>
                    ))
                } 
            </div>}
        </>
    )
}

GifGrid.propTypes = {
    categoria: PropTypes.string.isRequired,
    quantity: PropTypes.number.isRequired    
}
