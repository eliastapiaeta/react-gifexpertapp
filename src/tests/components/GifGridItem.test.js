import { shallow } from "enzyme";
import { GifGridItem } from "../../components/GifGridItem";

const dataImage = {
    id: 'JhZcAuGjuDmZq',
    title: 'happy amy poehler GIF by Disney',
    url: 'https://media0.giphy.com/media/JhZcAuGjuDmZq/giphy.gif?cid=8d830538qq68sxtbhmrpm2f49dg18f0m2svlora5xr35t2f4&rid=giphy.gif',
}
const wrapper = shallow(<GifGridItem
    id={dataImage.id} title={dataImage.title} url={dataImage.url}
/>);


describe('Prueba en GrifGridItem.js', () => {

    test('should return object::tag<html>', () => {

        expect(wrapper).toMatchSnapshot();
    });
    test('debe de tener un pÃ¡rrafo con el title', () => {
        const p = wrapper.find('p');
        expect(p.text().trim()).toBe(dataImage.title);
    });
    test('debe de tener la imagen al id, url y alt de los props', () => {
        const img = wrapper.find('img');

        // console.log(img.props());
        expect(img.prop('src')).toBe(dataImage.url);
        expect(img.prop('alt')).toBe(dataImage.title);
        expect(img.prop('id')).toBe(dataImage.id);
    });
    test('should tener animate__fadeInLeft', () => {
        const claseABuscar = 'animate__fadeInLeft';
        const div = wrapper.find('div');
        const className = div.props().className;
        // const clases = className.split(' ');
        // const claseIndice = clases.indexOf(claseABuscar);
        // const clase = clases[claseIndice];
        
        expect(className.includes(claseABuscar)).toBe(true);
    })
    

})
