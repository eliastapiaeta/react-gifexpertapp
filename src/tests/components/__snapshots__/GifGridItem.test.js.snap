// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en GrifGridItem.js should return object::tag<html> 1`] = `
<div
  className="card animate__animated animate__fadeInLeft"
>
  <img
    alt="happy amy poehler GIF by Disney"
    id="JhZcAuGjuDmZq"
    src="https://media0.giphy.com/media/JhZcAuGjuDmZq/giphy.gif?cid=8d830538qq68sxtbhmrpm2f49dg18f0m2svlora5xr35t2f4&rid=giphy.gif"
  />
  <p>
    happy amy poehler GIF by Disney
  </p>
</div>
`;
